### ISSUE TITLE
Add diagnostic for unsupported PEP 604 expressions

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4820]
PEP 604 introduced the `|` form for union type annotations, but some legal union expressions are not supported at runtime.

```python
Foo = int | "str" # Runtime error
```

It would be helpful if pyright reported this error when an unsupported expression.

**Comment by erictraut** [comment]
Pyright already does this for type annotations (when deferred evaluation is not used):

```python
def func(a: int | "str"): # Pyright reports an error here
    ...
```

This should be extended to runtime expressions as well.

**Comment by erictraut** [comment]
Turns out there's no good way to do this for expressions that are not expected to be types. Pyright already emits an error when the expression is used in an annotation or some other context where a type expression is expected (e.g. in a TypeVar bound or a TypeAlias definition).
